Add oshash support

WithoutPants opened the pull request:
Adds an `oshash` column to the database and makes `checksum` nullable. Added check to the scenes table to ensure at least one of the two are not null.

Introduces two new settings `calculateMD5` and `useMD5`. 

`calculateMD5` indicates whether or not to calculate the MD5 checksum during scanning. oshash will always be calculated for scenes where it is not already present. Similarly, if `calculateMD5` is true and the checksum is empty, then the MD5 will be calculated. 

`useMD5` indicates whether to use the MD5 checksum or the oshash for generated file naming and import/export naming.

For existing systems, these settings are both defaulted to true after migration if the scene table is not empty. 

New systems have `calculateMD5` and `useMD5` set to false.

The system prevents changing `useMD5` to false if there are any missing oshash values. Similarly, changing `useMD5` to true will be prevented if there are any scenes without checksums. Finally, `calculateMD5` is prevented from being set to false if `useMD5` is true.

![image](https://user-images.githubusercontent.com/53250216/87395836-51e76500-c5f5-11ea-880e-8ac10bd1d12f.png)

Adds `hash` to the file info panel in the UI. Missing hashes are not displayed.

![image](https://user-images.githubusercontent.com/53250216/87395786-3e3bfe80-c5f5-11ea-9675-6a534aefdf58.png)

Adds a manual migration task that renames existing generated files (not including exported metadata files) to the current naming format.

![image](https://user-images.githubusercontent.com/53250216/87395986-8c510200-c5f5-11ea-8654-95924f486299.png)

The oshash algorithm has been verified manually using the sample files from the opensubtitles website.

Resolves #351

bnkai commented on pkg/utils/oshash.go:
```diff
@@ -0,0 +1,82 @@
+	sum += uint64(fileSize)
+
+	// output as hex
+	return fmt.Sprintf("%16x", sum), nil
```
oshash should be 0 padded to 16 digits so
```go
return fmt.Sprintf("%016x", sum), nil
```

WithoutPants commented:
I think it's probably worth extending this to include galleries as well.

bnkai commented:
My only concern for galleries would be how `unique` the hash is if we use zip files or files that can be <64k ( small zip file with a single image, or plain image in the future). I think it was mainly tested with video files.

WithoutPants commented:
Yeah, you're right. Might be better leaving it off.

bnkai commented on pkg/manager/task_migrate_hash.go:
```diff
@@ -0,0 +1,84 @@
+
+	oldPath = scenePaths.GetSpriteVttFilePath(oldHash)
+	newPath = scenePaths.GetSpriteVttFilePath(newHash)
+	t.migrate(oldPath, newPath)
```
the actual sprite image is missing

```go
	oldPath = scenePaths.GetSpriteImageFilePath(oldHash)
	newPath = scenePaths.GetSpriteImageFilePath(newHash)
	t.migrate(oldPath, newPath)
```

bnkai commented on pkg/manager/task_migrate_hash.go:
```diff
@@ -0,0 +1,84 @@
+	newExists, _ := utils.FileExists(newName)
+
+	if oldExists {
+		if newExists {
```
Is there a reason not to overwrite existing files ?
If the user changes the naming hash/checksum and forgets to migrate the names then during the next scan/content generation he will get files with both naming schemes. Doing a migrate task afterwards will only print the warning and leave files as they are.

WithoutPants replied:
Good point. Fixed.

bnkai commented on pkg/manager/manager_tasks.go:
```diff
@@ -626,6 +639,54 @@ func (s *singleton) Clean() {
+
+		scenes, err := qb.All()
+		if err != nil {
+			logger.Errorf("failed to fetch list of scenes for cleaning")
```
even if the task is using the clean state isn't it more helpful to print something like "migrating" or "file renaming" or something like that?

bnkai commented on pkg/manager/manager_tasks.go:
```diff
@@ -626,6 +639,54 @@ func (s *singleton) Clean() {
+			}
+
+			if scene == nil {
+				logger.Errorf("nil scene, skipping Clean")
```
same as above for `Clean` 
as a side note `Clean` doesn't have a status message so nothing is printed in the status when doing the name migration

bnkai commented:
It seems to work ok both with an empty/existing db.
Initial oshash scan for my exitsing ~4TB, 11K scenes (stored on local HDDs) took ~6 minutes.
The name migration took < 5s ( nvme ssd)

WithoutPants commented:
Merged in from develop branch. There was quite a bit of conflicting changes, so may need a retest.

Fixed up the logging and status stuff identified. Also fixed a nasty panic when running clean if a file is not accessible.

bnkai commented on pkg/manager/task_scan.go:
```diff
@@ -182,15 +223,33 @@ func (t *ScanTask) scanScene() {
+		sceneHash = checksum
+		scene, _ = qb.FindByChecksum(sceneHash)
+	} else {
+		// TODO scene = findbyoshash
```
is a 
```golang
scene = qb.FindByOSHash(sceneHash)
```
here enough to remove the TODO?

bnkai commented on pkg/manager/task_migrate_hash.go:
```diff
@@ -0,0 +1,86 @@
+func (t *MigrateHashTask) migrate(oldName, newName string) {
+	oldExists, err := utils.FileExists(oldName)
+	if err != nil {
+		logger.Errorf("Error checking existence of %s: %s", oldName, err.Error())
```
should we switch from error level to debug or warning here?
It pollutes the log with `Error checking existence of ...` for marker,webp and transcoded files ( for the majority of my scenes i don't have those generated). The specific files not existing shouldn't be an error

WithoutPants replied:
Should've only been logging due to other error. Fixed.

bnkai commented:
Retested seems ok.

bnkai commented:
I finished retesting. It wasnt an exhaustive test though, more of a check for regressions due to refactoring test. Looks ok.

For users with existing databases i think a step by step migration guide could be useful.
Eg.

1. Scan library ( updates the oshash for existing entries)
2. Untick UseMD5
3. Choose oshash as the naming scheme
4. Migrate
